jax/_src/custom_call.py

```diff
@@ -0,0 +1,117 @@
+# Copyright 2024 The JAX Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+from collections.abc import Iterable, Sequence
+from typing import Any
+
+import numpy as np
+
+from jax._src import core
+from jax._src import dtypes
+from jax._src.interpreters import mlir
+from jax._src.interpreters.mlir import ir
+
+
+def custom_call_lowering(
+  call_target_name: str,
+  api_version: int = 4,
+  backend_config: str | bytes | dict[str, ir.Attribute] | None = None,
+  result_types: Sequence[ir.Type] | None = None,
```

Comment:
We have an option of using `jax.ShapeDtypeStruct` here instead of MLIR types. Wdyt?

Reply:
I haven't used `ShapeDtypeStruct` before so I'm not sure I totally understand what this would mean here? What are the benefits?

Reply:
The benefit is that the caller doesn't reach into the `ir` module to construct MLIR types and can use JAX's dtypes instead.

So, e.g. instead of 

```
result_types=[ir.RankedTensorType([42], ir.IntegerType.get(...)]
result_shapes=[42]
```

the caller can do 

```
result_shapes_dtypes=[jax.ShapeDtypeStruct([42], jnp.float32)]
```

Reply:
Got it! In thinking through this, I'm actually inclined to just remove `result_types` and `result_shapes` and not add `result_shapes_dtypes` either. If a user wants override either of these, this helper function really doesn't add much benefit over just calling `mlir.custom_call` directly.

It is still possible to customize these values using `**lowering_args`, but I don't think that they need to be exposed as top level parameters!